File: internal/upstream/balancer.go
package upstream

import (
	"math/rand/v2"
	"sync/atomic"
)

// Selector chooses which upstream address a request should be forwarded to.
//
// Next returns the chosen address and a token the caller must hand back to
// Release when the request completes. For load-aware selectors (phase 5:
// least-connections, power-of-two-choices) Release frees the upstream's
// in-flight slot; for the stateless selectors (Fixed, RoundRobin) the token
// is -1 and Release is a no-op. The token is a plain int (the upstream
// index) rather than a closure so selection stays allocation-free on the hot
// path.
type Selector interface {
	Next() (addr string, token int)
	Release(token int)
}

// Load-balancing strategy identifiers, selected via the LB_STRATEGY env var
// (phase 5). An empty strategy keeps the pre-phase-5 behavior (Fixed /
// plain RoundRobin), so phases 0-3 are unaffected.
const (
	LBRoundRobin = "round-robin" // phase 5a
	LBLeastConn  = "least-conn"  // phase 5b
	LBP2C        = "p2c"         // phase 5c
)

// Fixed is a Selector that always returns the same address (phase 0: a
// single upstream).
type Fixed string

// Next implements Selector.
func (f Fixed) Next() (string, int) { return string(f), -1 }

// Release implements Selector (no-op: Fixed tracks no load).
func (Fixed) Release(int) {}

// RoundRobin cycles through a fixed set of upstream addresses using an
// atomic counter, avoiding mutex contention between concurrent callers
// (phase 1a-3b: multiple upstreams, no load-awareness). It tracks no
// per-upstream state; phase 5's CountingRoundRobin adds that.
type RoundRobin struct {
	addrs   []string
	counter atomic.Uint64
}

// NewRoundRobin returns a RoundRobin selector over addrs. addrs must be
// non-empty.
func NewRoundRobin(addrs []string) *RoundRobin {
	return &RoundRobin{addrs: addrs}
}

// Next implements Selector.
func (r *RoundRobin) Next() (string, int) {
	i := r.counter.Add(1) - 1
	return r.addrs[i%uint64(len(r.addrs))], -1
}

// Release implements Selector (no-op: RoundRobin tracks no load).
func (*RoundRobin) Release(int) {}

// UpstreamStat is a snapshot of one upstream's routing counters, reported by
// selectors that implement StatsProvider (phase 5 distribution tracking).
type UpstreamStat struct {
	Addr     string `json:"addr"`
	Total    int64  `json:"total"`    // cumulative requests routed here
	InFlight int64  `json:"inflight"` // requests currently outstanding
}

// StatsProvider is implemented by the load-aware selectors so cmd/proxy can
// expose per-upstream request distribution (phase 5): so we can see where
// requests go, and whether the slow upstreams are being steered around.
type StatsProvider interface {
	// Stats returns a snapshot of every upstream's counters. If reset is
	// true, the cumulative Total counters are zeroed after the snapshot is
	// taken (used to isolate a measured run from its warmup); InFlight is a
	// live gauge and is never reset.
	Stats(reset bool) []UpstreamStat
}

// counters holds the shared per-upstream state the phase-5 selectors track,
// index-aligned to addrs. total is cumulative (for distribution reporting);
// inflight is a gauge incremented at selection and decremented at Release
// (the signal least-connections / P2C balance on).
type counters struct {
	addrs    []string
	total    []atomic.Int64
	inflight []atomic.Int64
}

func newCounters(addrs []string) counters {
	return counters{
		addrs:    addrs,
		total:    make([]atomic.Int64, len(addrs)),
		inflight: make([]atomic.Int64, len(addrs)),
	}
}

// pick records a selection on the upstream at idx and returns its address
// and idx as the release token.
func (c *counters) pick(idx int) (string, int) {
	c.total[idx].Add(1)
	c.inflight[idx].Add(1)
	return c.addrs[idx], idx
}

// Release implements Selector for all counting selectors.
func (c *counters) Release(token int) {
	if token >= 0 {
		c.inflight[token].Add(-1)
	}
}

// Stats implements StatsProvider for all counting selectors.
func (c *counters) Stats(reset bool) []UpstreamStat {
	stats := make([]UpstreamStat, len(c.addrs))
	for i := range c.addrs {
		total := c.total[i].Load()
		if reset {
			c.total[i].Add(-total)
		}
		stats[i] = UpstreamStat{
			Addr:     c.addrs[i],
			Total:    total,
			InFlight: c.inflight[i].Load(),
		}
	}
	return stats
}

// CountingRoundRobin is round-robin (phase 5a) plus the per-upstream
// counters, so its even distribution can be compared against the load-aware
// strategies. Selection is identical to RoundRobin; only the bookkeeping
// differs.
type CountingRoundRobin struct {
	counters
	counter atomic.Uint64
}

// NewCountingRoundRobin returns a CountingRoundRobin over addrs.
func NewCountingRoundRobin(addrs []string) *CountingRoundRobin {
	return &CountingRoundRobin{counters: newCounters(addrs)}
}

// Next implements Selector.
func (r *CountingRoundRobin) Next() (string, int) {
	i := r.counter.Add(1) - 1
	return r.pick(int(i % uint64(len(r.addrs))))
}

// LeastConn selects the upstream with the fewest in-flight requests (phase
// 5b), so slow upstreams — which accumulate in-flight requests while they
// sleep — receive proportionally less traffic. Ties (e.g. several idle fast
// upstreams all at zero) are broken by a uniform random choice among the
// tied minima via reservoir sampling, so load spreads evenly instead of
// piling onto the lowest index.
type LeastConn struct {
	counters
}

// NewLeastConn returns a LeastConn selector over addrs.
func NewLeastConn(addrs []string) *LeastConn {
	return &LeastConn{counters: newCounters(addrs)}
}

// Next implements Selector.
func (l *LeastConn) Next() (string, int) {
	best := 0
	bestLoad := l.inflight[0].Load()
	ties := 1
	for i := 1; i < len(l.addrs); i++ {
		load := l.inflight[i].Load()
		switch {
		case load < bestLoad:
			best, bestLoad, ties = i, load, 1
		case load == bestLoad:
			// Reservoir sample: pick each equal-minimum upstream with
			// probability 1/ties so ties spread uniformly.
			ties++
			if rand.IntN(ties) == 0 {
				best = i
			}
		}
	}
	return l.pick(best)
}

// P2C is power-of-two-choices (phase 5c): sample two distinct upstreams at
// random and route to the less loaded of the two. It approximates
// least-connections' load-shedding without the O(n) scan, and avoids the
// herd behavior a global-minimum selector can exhibit at high fan-out.
type P2C struct {
	counters
}

// NewP2C returns a P2C selector over addrs.
func NewP2C(addrs []string) *P2C {
	return &P2C{counters: newCounters(addrs)}
}

// Next implements Selector.
func (p *P2C) Next() (string, int) {
	n := len(p.addrs)
	i := rand.IntN(n)
	j := rand.IntN(n - 1)
	if j >= i {
		j++ // map j into [0,n) \ {i} so the two picks are distinct
	}
	if p.inflight[j].Load() < p.inflight[i].Load() {
		i = j
	}
	return p.pick(i)
}

// NewSelector returns a Selector for addrs under the given load-balancing
// strategy. An empty strategy preserves the pre-phase-5 behavior (Fixed for
// a single address, plain RoundRobin otherwise) so phases 0-3b are
// unchanged; the phase-5 strategies use the counting, load-aware selectors.
// An unrecognized non-empty strategy falls back to the pre-phase-5 behavior.
func NewSelector(addrs []string, strategy string) Selector {
	switch strategy {
	case LBRoundRobin:
		return NewCountingRoundRobin(addrs)
	case LBLeastConn:
		return NewLeastConn(addrs)
	case LBP2C:
		return NewP2C(addrs)
	}
	if len(addrs) == 1 {
		return Fixed(addrs[0])
	}
	return NewRoundRobin(addrs)
}

File: crates/service-proxy/src/upstream/balancer.rs
//! Load-balancing selectors. Mirrors go/internal/upstream/balancer.go.

use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
use std::sync::Arc;

use rand::RngExt;
use serde::Serialize;

pub const LB_ROUND_ROBIN: &str = "round-robin";
pub const LB_LEAST_CONN: &str = "least-conn";
pub const LB_P2C: &str = "p2c";

pub trait Selector: Send + Sync {
    fn next(&self) -> (String, i32);
    fn release(&self, token: i32);
}

pub trait StatsProvider: Selector {
    fn stats(&self, reset: bool) -> Vec<UpstreamStat>;
}

#[derive(Debug, Clone, Serialize)]
pub struct UpstreamStat {
    pub addr: String,
    pub total: i64,
    pub inflight: i64,
}

pub struct Fixed {
    addr: String,
}

impl Fixed {
    pub fn new(addr: String) -> Self {
        Self { addr }
    }
}

impl Selector for Fixed {
    fn next(&self) -> (String, i32) {
        (self.addr.clone(), -1)
    }
    fn release(&self, _token: i32) {}
}

pub struct RoundRobin {
    addrs: Vec<String>,
    counter: AtomicU64,
}

impl RoundRobin {
    pub fn new(addrs: Vec<String>) -> Self {
        Self {
            addrs,
            counter: AtomicU64::new(0),
        }
    }
}

impl Selector for RoundRobin {
    fn next(&self) -> (String, i32) {
        let i = self.counter.fetch_add(1, Ordering::Relaxed);
        let addr = self.addrs[i as usize % self.addrs.len()].clone();
        (addr, -1)
    }
    fn release(&self, _token: i32) {}
}

struct Counters {
    addrs: Vec<String>,
    total: Vec<AtomicI64>,
    inflight: Vec<AtomicI64>,
}

impl Counters {
    fn new(addrs: Vec<String>) -> Self {
        let n = addrs.len();
        Self {
            addrs,
            total: (0..n).map(|_| AtomicI64::new(0)).collect(),
            inflight: (0..n).map(|_| AtomicI64::new(0)).collect(),
        }
    }

    fn pick(&self, idx: usize) -> (String, i32) {
        self.total[idx].fetch_add(1, Ordering::Relaxed);
        self.inflight[idx].fetch_add(1, Ordering::Relaxed);
        (self.addrs[idx].clone(), idx as i32)
    }

    fn release(&self, token: i32) {
        if token >= 0 {
            self.inflight[token as usize].fetch_sub(1, Ordering::Relaxed);
        }
    }

    fn stats(&self, reset: bool) -> Vec<UpstreamStat> {
        self.addrs
            .iter()
            .enumerate()
            .map(|(i, addr)| {
                let total = self.total[i].load(Ordering::Relaxed);
                if reset {
                    self.total[i].fetch_sub(total, Ordering::Relaxed);
                }
                UpstreamStat {
                    addr: addr.clone(),
                    total,
                    inflight: self.inflight[i].load(Ordering::Relaxed),
                }
            })
            .collect()
    }
}

pub struct CountingRoundRobin {
    counters: Counters,
    counter: AtomicU64,
}

impl CountingRoundRobin {
    pub fn new(addrs: Vec<String>) -> Self {
        Self {
            counters: Counters::new(addrs),
            counter: AtomicU64::new(0),
        }
    }
}

impl Selector for CountingRoundRobin {
    fn next(&self) -> (String, i32) {
        let i = self.counter.fetch_add(1, Ordering::Relaxed);
        let idx = i as usize % self.counters.addrs.len();
        self.counters.pick(idx)
    }
    fn release(&self, token: i32) {
        self.counters.release(token);
    }
}

impl StatsProvider for CountingRoundRobin {
    fn stats(&self, reset: bool) -> Vec<UpstreamStat> {
        self.counters.stats(reset)
    }
}

pub struct LeastConn {
    counters: Counters,
}

impl LeastConn {
    pub fn new(addrs: Vec<String>) -> Self {
        Self {
            counters: Counters::new(addrs),
        }
    }

    /// Test helper: set inflight for index.
    #[cfg(test)]
    pub fn set_inflight(&self, idx: usize, v: i64) {
        self.counters.inflight[idx].store(v, Ordering::Relaxed);
    }
}

impl Selector for LeastConn {
    fn next(&self) -> (String, i32) {
        let mut best = 0usize;
        let mut best_load = self.counters.inflight[0].load(Ordering::Relaxed);
        let mut ties = 1usize;
        let mut rng = rand::rng();
        for i in 1..self.counters.addrs.len() {
            let load = self.counters.inflight[i].load(Ordering::Relaxed);
            if load < best_load {
                best = i;
                best_load = load;
                ties = 1;
            } else if load == best_load {
                ties += 1;
                if rng.random_range(0..ties) == 0 {
                    best = i;
                }
            }
        }
        self.counters.pick(best)
    }
    fn release(&self, token: i32) {
        self.counters.release(token);
    }
}

impl StatsProvider for LeastConn {
    fn stats(&self, reset: bool) -> Vec<UpstreamStat> {
        self.counters.stats(reset)
    }
}

pub struct P2C {
    counters: Counters,
}

impl P2C {
    pub fn new(addrs: Vec<String>) -> Self {
        Self {
            counters: Counters::new(addrs),
        }
    }

    #[cfg(test)]
    pub fn set_inflight(&self, idx: usize, v: i64) {
        self.counters.inflight[idx].store(v, Ordering::Relaxed);
    }
}

impl Selector for P2C {
    fn next(&self) -> (String, i32) {
        let n = self.counters.addrs.len();
        let mut rng = rand::rng();
        let mut i = rng.random_range(0..n);
        let mut j = rng.random_range(0..n - 1);
        if j >= i {
            j += 1;
        }
        if self.counters.inflight[j].load(Ordering::Relaxed)
            < self.counters.inflight[i].load(Ordering::Relaxed)
        {
            i = j;
        }
        self.counters.pick(i)
    }
    fn release(&self, token: i32) {
        self.counters.release(token);
    }
}

impl StatsProvider for P2C {
    fn stats(&self, reset: bool) -> Vec<UpstreamStat> {
        self.counters.stats(reset)
    }
}

/// Type-erased selector with optional stats.
pub enum SelectorKind {
    Fixed(Fixed),
    RoundRobin(RoundRobin),
    CountingRoundRobin(CountingRoundRobin),
    LeastConn(LeastConn),
    P2C(P2C),
}

impl Selector for SelectorKind {
    fn next(&self) -> (String, i32) {
        match self {
            Self::Fixed(s) => s.next(),
            Self::RoundRobin(s) => s.next(),
            Self::CountingRoundRobin(s) => s.next(),
            Self::LeastConn(s) => s.next(),
            Self::P2C(s) => s.next(),
        }
    }
    fn release(&self, token: i32) {
        match self {
            Self::Fixed(s) => s.release(token),
            Self::RoundRobin(s) => s.release(token),
            Self::CountingRoundRobin(s) => s.release(token),
            Self::LeastConn(s) => s.release(token),
            Self::P2C(s) => s.release(token),
        }
    }
}

impl SelectorKind {
    pub fn stats(&self, reset: bool) -> Option<Vec<UpstreamStat>> {
        match self {
            Self::CountingRoundRobin(s) => Some(s.stats(reset)),
            Self::LeastConn(s) => Some(s.stats(reset)),
            Self::P2C(s) => Some(s.stats(reset)),
            _ => None,
        }
    }

    pub fn has_stats(&self) -> bool {
        matches!(
            self,
            Self::CountingRoundRobin(_) | Self::LeastConn(_) | Self::P2C(_)
        )
    }
}

pub fn new_selector(addrs: Vec<String>, strategy: &str) -> Arc<SelectorKind> {
    let kind = match strategy {
        LB_ROUND_ROBIN => SelectorKind::CountingRoundRobin(CountingRoundRobin::new(addrs)),
        LB_LEAST_CONN => SelectorKind::LeastConn(LeastConn::new(addrs)),
        LB_P2C => SelectorKind::P2C(P2C::new(addrs)),
        _ if addrs.len() == 1 => SelectorKind::Fixed(Fixed::new(addrs.into_iter().next().unwrap())),
        _ => SelectorKind::RoundRobin(RoundRobin::new(addrs)),
    };
    Arc::new(kind)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn least_conn_picks_least_loaded() {
        let lc = LeastConn::new(vec!["a".into(), "b".into(), "c".into()]);
        lc.set_inflight(0, 10);
        lc.set_inflight(2, 3);
        let (addr, tok) = lc.next();
        assert_eq!(addr, "b");
        assert_eq!(tok, 1);
        lc.release(tok);
    }

    #[test]
    fn p2c_picks_less_loaded() {
        let p = P2C::new(vec!["x".into(), "y".into()]);
        p.set_inflight(0, 5);
        for _ in 0..20 {
            let (addr, tok) = p.next();
            assert_eq!(addr, "y");
            p.release(tok);
        }
    }

    #[test]
    fn counting_rr_even() {
        let rr = CountingRoundRobin::new(vec!["a".into(), "b".into(), "c".into()]);
        for _ in 0..30 {
            let (_, tok) = rr.next();
            rr.release(tok);
        }
        for s in rr.stats(false) {
            assert_eq!(s.total, 10);
            assert_eq!(s.inflight, 0);
        }
    }

    #[test]
    fn stats_reset() {
        let lc = LeastConn::new(vec!["a".into(), "b".into()]);
        lc.next();
        lc.next();
        let total: i64 = lc.stats(true).iter().map(|s| s.total).sum();
        assert_eq!(total, 2);
        let snap = lc.stats(false);
        assert_eq!(snap.iter().map(|s| s.total).sum::<i64>(), 0);
        assert_eq!(snap.iter().map(|s| s.inflight).sum::<i64>(), 2);
    }

    #[test]
    fn new_selector_dispatch() {
        assert!(matches!(
            &*new_selector(vec!["a".into()], ""),
            SelectorKind::Fixed(_)
        ));
        assert!(matches!(
            &*new_selector(vec!["a".into(), "b".into()], ""),
            SelectorKind::RoundRobin(_)
        ));
        assert!(new_selector(vec!["a".into(), "b".into()], LB_P2C).has_stats());
        assert!(!new_selector(vec!["a".into(), "b".into()], "").has_stats());
    }
}
